### PR TITLE
add Sumologic as backend

### PR DESCRIPTION
**Pull request by juju4**
this PR adds Sumologic log platform (https://www.sumologic.com/) driver with existing python sumologic-sdk

### CODE REVIEW

**Comment by petebryan** [comment]
Thanks very much for the contribution @juju4 I will take a look at this!

**Comment by petebryan** [comment]
//azp run

**Comment by ianhelle** [comment]
@juju4 - this is awesome! Full test coverage and beautiful code. 🥇

File: msticpy/data/drivers/sumologic_driver.py [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+            raise MsticpyConnectionError(
+                f"Error connecting to Sumologic: {err}",
+                title="Sumologic connection",
+                help_uri="https://msticpy.readthedocs.io/en/latest/DataProviders.html",
+            )
+        self._connected = True
+        print("connected")
+
+    def _get_connect_args(
+        self, connection_str: Optional[str], **kwargs
```
Comment (petebryan) [inline msticpy/data/drivers/sumologic_driver.py:115]: I can't see where connection_str is used in this function?

File: msticpy/data/drivers/sumologic_driver.py [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+        """
+        del query_source
+        if not self._connected:
+            raise self._create_not_connected_err()
+
+        verbosity = kwargs.pop("verbosity", 0)
+        timeZone = kwargs.pop("timeZone", 'UTC')
+        byReceiptTime = kwargs.pop("byReceiptTime", False)
+        forceMessagesResults = kwargs.pop("forceMessagesResults", False)
+        timeout = kwargs.pop("timeout", 300)
```
Comment (petebryan) [inline msticpy/data/drivers/sumologic_driver.py:184]: I don't see timeout used anywhere in this function - self.timeout is referenced however.

**Review by petebryan — COMMENTED** [review]
Thanks for the contribution - the functionality looks great, I just have a few comments. In addition can we please ask you run the linting and formatting we use in this project. Details can be found here (https://github.com/microsoft/msticpy/wiki/Contributor-guidelines) but please let us know if you have any issues with these and we can help.

File: msticpy/data/drivers/sumologic_driver.py [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+        if status['state'] == 'DONE GATHERING RESULTS' and (
+                not re.search(r'\|\s*count', query, re.IGNORECASE) or forceMessagesResults
+                ):
+            # Non-aggregated results, Messages only
+            count = status['messageCount']
+            limit2 = count if count < limit and count != 0 else limit # compensate bad limit check
+            try:
+                r = self.service.search_job_messages(sj, limit=limit2)
+                return(r['messages'])
+            except Exception as e:
```
[the remark this answers is not in the document]
Comment (ianhelle) [reply msticpy/data/drivers/sumologic_driver.py:246]: Also, can you avoid using 1 or 2 letter var names? one of our linters flags this as an issue. Often it's easy to see what it's for when its use is very localized but it's always nicer to have something more explicit - result, col, etc.

File: msticpy/data/drivers/sumologic_driver.py [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+
+        status = self.service.search_job_status(sj)
+        if verbosity >=2:
+            print("DEBUG: status {0}".format(status))
+        while status['state'] != 'DONE GATHERING RESULTS':
+            if status['state'] == 'CANCELLED':
+                break
+            status = self.service.search_job_status(sj)
+            if verbosity >=4:
+                print("DEBUG: pending results, state {0}. sleeping {1}}".format(status,self.timeout))
```
[the remark this answers is not in the document]
Comment (juju4) [reply msticpy/data/drivers/sumologic_driver.py:229]: yes. fixed

File: msticpy/data/drivers/sumologic_driver.py [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+        if 'limit' in kwargs:
+            query += "| limit {0}".format(kwargs['limit'])
+        limit = kwargs.pop("limit", 10000)
+
+        if verbosity >=1:
+            print("INFO: from {0} to {1}, timezone {2}".format(start_time, end_time, timeZone))
+        if verbosity >=2:
+            print("DEBUG: query {0}".format(query))
+            print("DEBUG: byReceiptTime {0}".format(byReceiptTime))
+            timer_start = timer()
```
[the remark this answers is not in the document]
Comment (juju4) [reply msticpy/data/drivers/sumologic_driver.py:212]: likely lost in a merge.
from timeit.
I'm putting under if as don't think good to import all the time when not used. any preferences?

File: msticpy/data/drivers/sumologic_driver.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (juju4) [reply msticpy/data/drivers/sumologic_driver.py:246]: for variables and exception granularity, that may need a bit more work. will look into.

File: msticpy/data/drivers/sumologic_driver.py [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+
+        if verbosity >=3:
+            print("DEBUG: {0}".format(results))
+        if normalize:
+            df = pd.json_normalize(results)
+        else:
+            df = pd.DataFrame(results)
+
+        for c in df.columns:
+            if c == "map._count" or c == "map._timeslice":
```
[the remark this answers is not in the document]
Comment (juju4) [reply msticpy/data/drivers/sumologic_driver.py:320]: fixed

File: msticpy/data/drivers/sumologic_driver.py [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+#  --------------------------------------------------------------------------
+"""Sumologic Driver class."""
+from datetime import datetime,timedelta
+import re
+import sys
+import time
+from typing import Any, Tuple, Union, Dict, Iterable, Optional
+
+import pandas as pd
+from sumologic import SumoLogic
```
[the remark this answers is not in the document]
Comment (juju4) [reply msticpy/data/drivers/sumologic_driver.py:14]: strange. working on my side but change done

File: msticpy/data/drivers/sumologic_driver.py (hunk 1 above)
Comment (juju4) [reply msticpy/data/drivers/sumologic_driver.py:115]: likely inherited from other driver template and to be removed

File: msticpy/data/drivers/sumologic_driver.py (hunk 2 above)
Comment (juju4) [reply msticpy/data/drivers/sumologic_driver.py:184]: fixed. self.timeout

File: msticpy/data/drivers/sumologic_driver.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (juju4) [reply msticpy/data/drivers/sumologic_driver.py]: for exception, added HTTPError and use msticpy usual raise
did also change variable names and more pylint
Likely will need a bit of time to test

File: msticpy/data/drivers/sumologic_driver.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (petebryan) [reply msticpy/data/drivers/sumologic_driver.py:209]: That sounds good, the more precise we can be with imports the better.

**Comment by juju4** [comment]
anything else needed to get this merged? thanks

**Comment by ianhelle** [comment]
@juju4 - thanks for the updates. I did see your commits happening but figured I'd wait for a message.
It will prob need a bit of work to merge to current main branch but I can do this. Thanks for the update.

File: msticpy/data/drivers/sumologic_driver.py [hunk 8]
```diff
@@ -0,0 +1,541 @@
...
+            byReceiptTime:  if time reference should used _receiptTime (time when Sumologic
+                        got message) instead of _messageTime (time present in log message).
+            limit: An integer describing the max number of search results to return.
+            forceMessagesResults: Force results to be raw messages even if aggregated query.
+            verbosity: Provide more verbose state. from 0 least verbose to 4 most one.
+            normalize: If set to True, fields containing structures (i.e. subfields)
+                   will be flattened such that each field has it's own column in
+                   the dataframe. If False, there will be a single column for the
+                   structure, with a JSON string encoding all the contents.
+            exporting: Export result to file.
```
Comment (ianhelle) [inline msticpy/data/drivers/sumologic_driver.py:353]: Do you need an "exporting" flag as well as "export_path"? If export_path is None then just ignore the export part

File: requirements.txt [hunk 9]
```diff
@@ -43,6 +43,7 @@ requests>=2.21.1
 setuptools>=40.6.3
 # splunk-sdk>=1.6.0  # splunk
 # statsmodels>=0.11.1  # timeseries
+sumologic-sdk>=0.1.11
```
Comment (ianhelle) [inline requirements.txt:46]: I moved this to an extra in setup.py

File: tests/data/drivers/test_sumologic_driver.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
+# -------------------------------------------------------------------------
```
Comment (ianhelle) [inline tests/data/drivers/test_sumologic_driver.py:1]: See the branch that I submitted. I couldn't get any of these tests to work. Seems like it was based on the Splunk driver test code and has a lot of unrelated stuff from that.

File: tests/data/drivers/test_sumologic_driver.py [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+SUMOLOGIC_CLI_PATCH = SumologicDriver.__module__ + ".sp_client"
+SUMOLOGIC_RESULTS_PATCH = SumologicDriver.__module__ + ".sp_results"
+
+
+# pylint: disable=too-many-branches, too-many-return-statements
+# pylint: disable=no-self-use, too-few-public-methods, protected-access
+
+
+def cli_connect(**kwargs):
```
Comment (ianhelle) [inline tests/data/drivers/test_sumologic_driver.py:35]: I moved a version of this into the _MockSumoLogicService class in my branch

File: tests/data/drivers/test_sumologic_driver.py [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+        return "other"
+
+
+class _MockAlert:
+    def __init__(self, name, count):
+        self.name = name
+        self.count = count
+
+
+class _MockSumologicService(MagicMock):
```
Comment (ianhelle) [inline tests/data/drivers/test_sumologic_driver.py:73]: See the updates I did (partially) in the new branch.
The mock has to replicate the SumoLogic class otherwise it isn't doing it's job.
You need to implement search_job, search_job_status and search_job_records.
- search_job - this should take the params and return an emulated job object
- search_job_status - returns different results (you can drive the expected result status by setting _MockSumologicService.status to the desired value.
- search_job_records - should return json objects emulating SL results and check that they are correctly converted into the expect DF.

File: tests/data/drivers/test_sumologic_driver.py [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+            _MockAlert("alert4", 10),
+        ]
+
+    @staticmethod
+    def _query_response(query, **kwargs):
+        del kwargs
+        return query
+
+
+def _results_reader(query_result):
```
Comment (ianhelle) [inline tests/data/drivers/test_sumologic_driver.py:107]: not needed

File: tests/data/drivers/test_sumologic_driver.py [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+    """Mock Sumologic results reader."""
+    if "zero query" in query_result:
+        yield None
+    else:
+        for i in range(10):
+            yield {"row": i, "query": query_result, "text": f"test text {i}"}
+
+
+@patch(SUMOLOGIC_CLI_PATCH)
```
Comment (ianhelle) [inline tests/data/drivers/test_sumologic_driver.py:116]: Fixed patch in new branch

File: tests/data/drivers/test_sumologic_driver.py [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+    with pytest.raises(MsticpyConnectionError) as mp_ex:
+        sumologic_driver.connect(url="HTTPError", accessid="***", accesskey="***")  # nosec
+        print("connected", sumologic_driver.connected)
+        check.is_false(sumologic_driver.connected)
+    check.is_in("Sumologic connection", mp_ex.value.args)
+
+
+@patch(SUMOLOGIC_RESULTS_PATCH)
+@patch(SUMOLOGIC_CLI_PATCH)
+def test_sumologic_query_success(sumologic_client, sumologic_results):
```
Comment (ianhelle) [inline tests/data/drivers/test_sumologic_driver.py:168]: This doesn't work at all since the Mock is not correctly implemented.

**Review by ianhelle — CHANGES_REQUESTED** [review]

**Comment by ianhelle** [comment]
@juju4
I've pushed a new branch with the changes I mentioned in the review. https://github.com/microsoft/msticpy/tree/pr/116-sumologic-driver-2021-04-21.
This is merged with master so should be up-to-date

**Comment by ianhelle** [comment]
@juju4 - since I'm really keen to help you get this finished, I've done the refactoring on the sumologic_driver class. I removed a couple methods/properties that were specific to Splunk (although we can add these back if there is an SL equivalent).
It really needs just the unit tests sorting out. Since I don't have access to a Sumologic subscription I think you'll need to test this against a live service.

Thx for doing this - it's very cool. My apologies for letting it run for so long - we were doing a lot of work for 1.0 release.

**Comment by juju4** [comment]
Thanks a lot @ianhelle 
yes, initial work was based on splunk implementation.

Will review and comment back  but on testing, whatever the backend, not sure how to handle it well and with CI as without a live server/active cloud subscription, can't really be tested...
not aware of also libraries that could help to validate queries in an offline way.

**Comment by ianhelle** [comment]
Merged with main branch via #165.
